### PR TITLE
Improve documentation for bitops

### LINKED ISSUE #7587: stdlib bitops incorrect result for popcount
```
echo popcount(0b11111111'i8)
```
returns 32 which can't be right given that byte has only 8 bits.

P.S.
Internally popcount converts ``int8`` to ``uint32`` which produces wrong result for for negative numbers due to sign extention. It needs to cast ``int8` to `uint8` first then convert to `uint32`  or implement `uint8` directly

### PR DESCRIPTION
**Pull request by konsumlamm**
* improve wording
* use `func` (instead of `proc` with `{.noSideEffect.}`)
* use `let` in `runnableExamples` where possible
* add some links

EDIT: Closes #7587.

### CODE REVIEW

**Comment by timotheecour** [comment]
let's wait for https://github.com/nim-lang/Nim/pull/16622 to get merged 1st as that PR has been in the queue for a while, to avoid conflicts there

**Comment by konsumlamm** [comment]
* all the tests in `main2` were already in `main1`, so i renamed `main1` to `main` and removed `main2`
* add a test as suggested in https://github.com/nim-lang/Nim/issues/7587#issuecomment-629617266

**Comment by timotheecour** [comment]
@konsumlamm ok https://github.com/nim-lang/Nim/pull/16622 got merged, please rebase :)

File: lib/pure/bitops.nim [hunk 1]
```diff
@@ -479,24 +483,24 @@ when useGCC_builtins:
 
 elif useVCC_builtins:
   # Counts the number of one bits (population count) in a 16-, 32-, or 64-byte unsigned integer.
-  proc builtin_popcnt16(a2: uint16): uint16 {.
-      importc: "__popcnt16", header: "<intrin.h>", noSideEffect.}
-  proc builtin_popcnt32(a2: uint32): uint32 {.
-      importc: "__popcnt", header: "<intrin.h>", noSideEffect.}
-  proc builtin_popcnt64(a2: uint64): uint64 {.
-      importc: "__popcnt64", header: "<intrin.h>", noSideEffect.}
+  func builtin_popcnt16(a2: uint16): uint16 {.
```
Comment (timotheecour) [inline lib/pure/bitops.nim:486]: optional: turn into more readable 1-liners?
```nim
func builtin_popcnt16(a2: uint16): uint16 {.importc: "__popcnt16", header: "<intrin.h>".}
```
(even if exceeds that (IMO silly) 80 col limit a bit)

File: lib/pure/bitops.nim [hunk 2]
```diff
@@ -562,13 +566,13 @@ proc countSetBits*(x: SomeInteger): int {.inline, noSideEffect.} =
...
-proc popcount*(x: SomeInteger): int {.inline, noSideEffect.} =
-  ## Alias for for `countSetBits <#countSetBits,SomeInteger>`_. (Hamming weight.)
+func popcount*(x: SomeInteger): int {.inline.} =
+  ## Alias for `countSetBits <#countSetBits,SomeInteger>`_ (Hamming weight).
   result = countSetBits(x)
 
-proc parityBits*(x: SomeInteger): int {.inline, noSideEffect.} =
-  ## Calculate the bit parity in integer. If number of 1-bit
-  ## is odd parity is 1, otherwise 0.
+func parityBits*(x: SomeInteger): int {.inline.} =
```
Comment (timotheecour) [inline lib/pure/bitops.nim:573]: pre-existing, but why isn't this returning `bool`? nim isn't C...; casting should have no cost, and resulting api would be easier; maybe we need another proc that returns bool (maybe `bitParity` ?)

File: lib/pure/bitops.nim [hunk 3]
```diff
@@ -589,10 +593,10 @@ proc parityBits*(x: SomeInteger): int {.inline, noSideEffect.} =
       when sizeof(x) <= 4: result = parityImpl(x.uint32)
       else: result = parityImpl(x.uint64)
 
-proc firstSetBit*(x: SomeInteger): int {.inline, noSideEffect.} =
-  ## Returns the 1-based index of the least significant set bit of x.
-  ## If `x` is zero, when ``noUndefinedBitOpts`` is set, result is 0,
-  ## otherwise result is undefined.
+func firstSetBit*(x: SomeInteger): int {.inline.} =
+  ## Returns the 1-based index of the least significant set bit of `x`.
+  ## If `x` is zero, when `noUndefinedBitOpts` is set, the result is 0,
```
Comment (timotheecour) [inline lib/pure/bitops.nim:598]: pre-existing but IMO future PR should add `-d:nimNoUndefinedBitOpts` of which noUndefinedBitOpts would become an undocumented alias

**Review by timotheecour — APPROVED** [review]
LGTM, I've checked everything including fact that large `main2` block was entirely covered by other tests block even though that's not obvious at 1st sight: some static blocks in main2 are covered by fact we have `static: main`

the duplication was introduced in https://github.com/nim-lang/Nim/pull/13803 (note that before https://github.com/nim-lang/Nim/pull/13803 bitops were also tested with 
-d:noIntrinsicsBitOpts
-d:noUndefinedBitOps
(see https://github.com/nim-lang/Nim/pull/13803/files#diff-cc11adc098ea503d26c0788e763173c275570c6074503ebea374ed348344562c)

in future PR we should add:
```nim
discard """
 matrix: "; -d:noIntrinsicsBitOpts -d:noUndefinedBitOps"
"""
```
to test this as well

File: lib/pure/bitops.nim (hunk 1 above)
Comment (timotheecour) [reply lib/pure/bitops.nim:486]: let's defer this to an eventual followup PR instead

**Pull request merged**